### PR TITLE
executor: remove *MVMap from agg executor

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve? 
Fix https://github.com/pingcap/tidb/issues/7450

- [x] CPU profile to be uploaded.

### What is changed and how it works?
Change groupMap of HashAggExec and HashAggFinalWorker from `*MVMap` to
`stringSet`, thus we can avoid some cost of `runtime.sliceToString`,
`runtime.byteSliceToString`, and the access to MVMap.

### Check List 

Tests 
Exist tests

 - No code

Code changes

 - Has exported function/method change

Side effects

Related changes

### CODE REVIEW

File: executor/aggregate.go [hunk 1]
```diff
@@ -19,15 +19,13 @@ import (
...
-	"github.com/pingcap/tidb/mysql"
 	"github.com/pingcap/tidb/sessionctx"
 	"github.com/pingcap/tidb/sessionctx/stmtctx"
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/util/chunk"
-	"github.com/pingcap/tidb/util/codec"
-	"github.com/pingcap/tidb/util/mvmap"
 	"github.com/spaolacci/murmur3"
 	"golang.org/x/net/context"
+	"sort"
```
Comment (shenli) [inline executor/aggregate.go]: Move this to the first block.

File: executor/aggregate.go [hunk 2]
```diff
@@ -379,26 +373,22 @@ func (w *HashAggPartialWorker) shuffleIntermData(sc *stmtctx.StatementContext, f
 }
 
 // getGroupKey evaluates the group items and args of aggregate functions.
-func (w *HashAggPartialWorker) getGroupKey(sc *stmtctx.StatementContext, row chunk.Row) ([]byte, error) {
-	w.groupValDatums = w.groupValDatums[:0]
+func (w *HashAggPartialWorker) getGroupKey(ctx sessionctx.Context, row chunk.Row) (groupKey string, _ error) {
 	for _, item := range w.groupByItems {
-		v, err := item.Eval(row)
+		v, isNull, err := item.EvalString(ctx, row)
```
Comment (zz-jason) [inline executor/aggregate.go]: we can not simply call `EvalString` here.

File: executor/aggregate.go [hunk 3]
```diff
@@ -621,22 +606,21 @@ func (e *HashAggExec) unparallelExec(ctx context.Context, chk *chunk.Chunk) erro
...
 	}
 	chk.Reset()
-	for {
-		groupKey, _ := e.groupIterator.Next()
-		if groupKey == nil {
-			return nil
-		}
-		partialResults := e.getPartialResults(groupKey)
+
+	sort.Strings(e.sortedGroupKey)
```
Comment (zz-jason) [inline executor/aggregate.go]: Why this sort is needed?

File: executor/aggregate.go [hunk 4]
```diff
@@ -680,35 +666,28 @@ func (e *HashAggExec) execute(ctx context.Context) (err error) {
...
 
-func (e *HashAggExec) getGroupKey(row chunk.Row) ([]byte, error) {
-	vals := make([]types.Datum, 0, len(e.GroupByItems))
+func (e *HashAggExec) getGroupKey(ctx sessionctx.Context, row chunk.Row) (groupKey string, _ error) {
 	for _, item := range e.GroupByItems {
-		v, err := item.Eval(row)
-		if item.GetType().Tp == mysql.TypeNewDecimal {
-			v.SetLength(0)
-		}
+		v, isNull, err := item.EvalString(ctx, row)
```
Comment (zz-jason) [inline executor/aggregate.go]: Maybe it's not safe to call `EvalString`?

**Comment by shenli** [comment]
@XuHuaiyu Please post the benchmark result here.

**Comment by coocood** [comment]
The `stringSet` has many pointers, GC overhead would be much higher.

**Comment by XuHuaiyu** [comment]
1. There is no obvious improvement for `T1` mentioned in https://github.com/pingcap/tidb/issues/7450#issuecomment-415244032:

master | this PR 
------ | --------
43.0901493788s | 42.8431748867s
[old.txt](https://github.com/pingcap/tidb/files/2346953/old.txt) | [new.txt](https://github.com/pingcap/tidb/files/2346954/new.txt)

Optimize points:
- [ ] ~3.77% `runtime.stringToByteSlice`of CPU cost introduced by `HashAggPartialWorker.shuffleIntermData`~
This cannot be avoid since we need to cast string as byte slice to calculate the hash value.
- [x] 0.75% `runtime.sliceToString` of CPU cost introduced by `baseHashAggWorker.getPartialResult`
- [x] 8.86% `MVMap.Get` and `MVMap.Put` of CPU cost introduced by access to MVMap

The performance is not improved obviously because that the access to StringSet which is used to replace MVMap seems cost as much as the access to MVMap.
But it would be **more readable** after using StringSet instead of MVMap.

2. For `T2` mentioned in: https://github.com/pingcap/tidb/issues/7450#issuecomment-415252169

After using a string slice to store the group keys and traverse it instead of the MVMap when fetching the final result, we got about **33.47%** CPU performance improvement.

master | this PR
------ | --------
463.533461905s | 308.472225189s
[old.txt](https://github.com/pingcap/tidb/files/2346955/old.txt) | [new.txt](https://github.com/pingcap/tidb/files/2346956/new.txt)

Optimize points:
- [x] 4.5% `runtime.makeslice` introduced by the definition of `vals` in HashAggExec.getGropuKey
- [x] 17.5% `MVMap.Get and MVMap.Put` introduced by the access to MVMap

**Comment by XuHuaiyu** [comment]
Seen from the profile picture, it seems the GC overhead is not much more than the master branch. @coocood

File: executor/aggregate.go [hunk 5]
```diff
@@ -680,35 +675,34 @@ func (e *HashAggExec) execute(ctx context.Context) (err error) {
 	}
 }
 
-func (e *HashAggExec) getGroupKey(row chunk.Row) ([]byte, error) {
-	vals := make([]types.Datum, 0, len(e.GroupByItems))
+func (e *HashAggExec) getGroupKey(row chunk.Row) (string, error) {
```
Comment (zz-jason) [inline executor/aggregate.go]: This function can be further optimized. Maybe in the next PR.

**Comment by zz-jason** [comment]
@XuHuaiyu `308.472225189s` is too slow, what is the corresponding query?

File: executor/aggregate.go (hunk 5 above)
Comment (zz-jason) [reply executor/aggregate.go]: we can change the way to calculate the `[]byte` according to the `group by` items.

Comment (lysu) [reply executor/aggregate.go]: this method maybe better return `[]byte` directly and make to string lazy happen, and then modify `StringSet` add a `Exists` override method that accept `[]byte` as argument, then use this small trick https://github.com/golang/go/commit/8072f46abdf5b2d3ed0ee7d691823b7fcdaa7c21#diff-590553aade2ea7ef319660a8151c5eb9R581

I feel @coocood want to make `StringSet` be `map[[n]byte]struct` if we have some "magic design" make groupKey be fixed length? :cat:

Comment (zz-jason) [reply executor/aggregate.go]: maybe `hack.String` or `hack.Slice` is what you want.

**Pull request merged**